ipatests: add test for kdcproxy handling reply split to several TCP packets

wladich opened the pull request:
This is a regression test for the bug in python-kdcproxy mentioned in
    https://github.com/latchset/kdcproxy/pull/44
      When the reply from AD is split into several TCP packets the kdc
      proxy software cannot handle it and returns a false error message
      indicating it cannot contact the KDC server.
    
    This could be observed as login failures of AD user on IPA clients
    when:
    * IPA client was configured to use kdcproxy to communicate with AD
    * kdcproxy used TCP to communicate with AD
    * response from AD to kdcproxy was split into several packets
    
    This patch also refactors and improves existing tests:
    * switch to using pytest fixtures for test setup and cleanup steps to make
      them isolated and reusable
    * simulate a much more restricted network environment: instead of blocking
      single 88 port we now block all outgoing traffic except few essential
      ports
    * add basic tests for using kdcproxy to communicate between IPA client
      and AD DC.

wladich commented:
Successful runs: [Fedora 32](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/ada90ec0-7ddd-11eb-a7dc-fa163e45d937/report.html), [Fedora 33](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/ef87f8ec-7ddd-11eb-9dd0-fa163e95bc24/report.html)
Removing temp commit

amore17 commented:
LGTM

amore17 commented on ipatests/test_integration/test_http_kdc_proxy.py:
```diff
@@ -4,53 +4,210 @@
+            paths.KRB5_CONF, encoding='utf-8')
+        kdc_url = 'https://{}/KdcProxy'.format(self.master.hostname)
+        kdc_option = 'kdc = {}'.format(kdc_url)
+
```
Also please add this for `kpasswd_server`

wladich replied:
Thank you for the comment. Reading about `kpaswd_server` option brought me to kdc proxy [design page](https://www.freeipa.org/page/V4/KDC_Proxy). There I have found the test plan which has few more checks that were implemented, one of them being test for kpasswd requests.

The PR has been updated :
* `kdc_option` added to krb5.conf file
* added checks for `kvno` and `kpasswd` commands according to the design page test plan.
* moved credentials cleanup to autouse fixture
* refactored `client_use_kdcproxy` fixture

amore17 replied:
Thanks @wladich LGTM

wladich commented:
The PR has been updated :

* kdc_option added to krb5.conf file
* added checks for kvno and kpasswd commands according to the [design page](https://www.freeipa.org/page/V4/KDC_Proxy) test plan.
* moved credentials cleanup to autouse fixture
* refactored client_use_kdcproxy fixture

wladich commented:
[Successful run](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/f82ae638-873f-11eb-85fa-fa163e45d937), removing temp commit

wladich commented:
master:

* 38c897f94640cc8ab64cb8afd128f98d62783379 ipatests: return result of kinit_as_user, pass raiseonerr parameter
* 40a686ca840f1f2ce4c4ebf6caa20b5947bae5f9 ipatests: add test for kdcproxy handling reply split to several TCP packets
* 76dd9a97fd633d7ee2d822962c8a106830b2e4aa ipatests: update prci definitions for test_http_kdc_proxy

wladich commented:
Backport to ipa-4-6 would require to also backport commits from https://pagure.io/freeipa/issue/7755. 
Removing ipa-4-6 label

stanislavlevin commented:
@wladich, hello.
This patch uncoditionally added dependency on AD to existed test class. I run this test on Azure Pipelines which hasn't support for AD(at least for now). I think that AD specifics can be moved out to its own test class.
